src/react-hook/react-hook.tsx

```diff
@@ -1,36 +1,22 @@
 import React from "react";
 import * as Config from "../module-config/module-config";
 
-// I was thinking I'd have the context shared by having the module
-// writer do
-// <UseConfig moduleName="@openmrs/esm-login">
-//   < ... >
-//     <Bar />
-//   </ ... >
-// </UseConfig>
-//
-// where Bar would ideally be able to do
-//   const moduleConfig = useModuleConfig()
-//
-type UseConfigProps = { moduleName: string; children: React.ReactNode };
-export function UseConfig(props: UseConfigProps) {
-  const ModuleNameContext = React.createContext(props.moduleName);
+export const ModuleNameContext = React.createContext(null);
 
-  // how do I make this available to descendants?
-  function useModuleConfig() {
-    const [config, setConfig] = React.useState(null);
-    const moduleName = React.useContext(ModuleNameContext);
-    if (!config) {
-      throw Config.getConfig(moduleName);
-    } else {
-      return config;
-    }
+export function useConfig() {
```

Comment:
Creating a typescript definition here could result in some great autocompletion for devs, and even **strictly typed configs with autocomplete** if we use generics.

Fancy genericized strict typed:

```ts
export function useConfig<T = object>(): T {
  // implementation
}

// Now people can use it with generics and get strict typing!
const config = useConfig<LoginConfig>()
config.logoUrl + 1 // Typescript will know that logoUrl is a string! And that adding 1 to it doesn't make sense

type LoginConfig = {
  logoUrl: string;
}
```

To get the autocompletion working in other repos, we'd have to set up a couple things in the `npm run build` script in this repo that aren't currently there. But adding a ts defn here sets us up for that.

Reply:
Awesome. Yeah, I was wondering if there might be a way to infer the type from the schema, since they should be homomorphic?

Reply:
Hmm maybe so, idk. We should check out if typescript has that feature. That would be very cool.